pigpen-cascading/src/main/clojure/pigpen/cascading.clj

```diff
@@ -41,335 +68,228 @@
+        new-fields (cfields (get-in projections [0 :alias]))]
+    (->
+      (PigPenAggregateBy. context pipe Fields/NONE old-fields)
+      (Rename. old-fields new-fields))))
+
+(s/defmethod command->flowdef :group
+  [{:keys [id keys fields join-types opts]} :- m/Group
+   ancestors
+   _]
+  (let [join-keys (group-key-cfields keys (:join-nils opts))
+        pipes (pipe-array
+                (if (:join-nils opts)
+                  (map :pipe ancestors)
+                  ;; This is to induce pigpen's nil-joining behavior, which
+                  ;; treats nils from the same relation as equal
+                  (->> ancestors
+                    (map-indexed
+                      (fn [i a]
+                        (let [fields (-> a :ancestor :fields cfields)]
+                          (Each. ^Pipe (:pipe a)
+                                 (InduceSentinelNils. (int i) fields))))))))]
+    (CoGroup. (str id) pipes join-keys Fields/NONE (BufferJoin.))))
+
+(s/defmethod command->flowdef :group-fold
+  [{:keys [group :- m/Group
+           fold :- m/Mapcat]}
+   ancestors
+   _]
+  (let [{:keys [id keys]} group
+        join-keys (group-key-cfields keys (get-in group [:opts :join-nils]))
+        ;; Perform all fold aggregations before co-grouping the relations
+        pipes (pipe-array
+                (zip [p (-> fold :projections next)
+                      {:keys [pipe]} ancestors
+                      key keys]
+                  (if (some-> p :expr :udf #{:fold})
+                    (let [context (pr-str `'{:projections ~(prepare-projections [p])})
+                          group-fields (cfields [key])
+                          arg-fields (cfields (get-in p [:expr :args]))]
+                      (PigPenAggregateBy. context pipe group-fields arg-fields))
+                    pipe)))]
+    (CoGroup. (str id) pipes join-keys Fields/NONE (BufferJoin.))))
+
+(s/defmethod command->flowdef :join
+  [{:keys [id keys fields join-types opts]} :- m/Join
+   ancestors
+   _]
+  (let [pipes (->> ancestors
+                (map :pipe)
+                (pipe-array))
+        joiner (->> join-types
+                 (map (comp boolean #{:required}))
+                 (boolean-array)
+                 (MixedJoin.))
+        join-keys (group-key-cfields keys (:join-nils opts))]
+    (CoGroup. (str id) pipes join-keys (cfields fields) joiner)))
+
+(s/defmethod command->flowdef :generate
+  [{:keys [id projections fields]} :- m/Mapcat
+   [{:keys [^Pipe pipe ancestor]}]
+   _]
+  (let [context (pr-str `'{:ancestor ~(update-in ancestor [:opts] dissoc :tap)
+                           :projections ~(prepare-projections projections)
+                           :fields ~fields})]
+    (case (:type ancestor)
+      :reduce
+      (Every. pipe (ReduceBuffer. context (cfields fields)) Fields/RESULTS)
+
+      (:group :group-fold)
+      (Every. pipe (GroupBuffer. context (cfields fields)) Fields/RESULTS)
```

Comment:
Related to the comment above, I think group-fold needs to be handled differently from group.